x/meshsecurity/contract/out_message.go

```diff
@@ -1,5 +1,27 @@
 package contract
 
-type SudoMsg struct {
-	Rebalance *struct{} `json:"rebalance"`
-}
+import (
+	wasmvmtypes "github.com/CosmWasm/wasmvm/types"
+)
+
+type (
+	SudoMsg struct {
+		Rebalance    *struct{}     `json:"rebalance,omitempty"`
+		ValsetUpdate *ValsetUpdate `json:"valset_update,omitempty"`
+	}
+
+	// Validator alias to wasmVM type
+	Validator = wasmvmtypes.Validator
+	// ValidatorAddr alias for the Bech32 address string of sdk.ValAddress
+	ValidatorAddr = string
+
+	// ValsetUpdate updates to the active validator set
+	ValsetUpdate struct {
+		Additions  []Validator     `json:"additions,omitempty"`
+		Removals   []ValidatorAddr `json:"removals,omitempty"`
+		Updated    []Validator     `json:"updated,omitempty"`
+		Jailed     []ValidatorAddr `json:"slashed,omitempty"`
+		Unjailed   []ValidatorAddr `json:"unjailed,omitempty"`
+		Tombstoned []ValidatorAddr `json:"tombstoned,omitempty"`
+	}
```

Comment:
👍🏼 Makes sense. Will update the message accordingly on the contract side.

Reply:
Here's how I understand it. Please correct it if wrong:

- `Additions`: New and active validators.
- `Removals`: Validators that leave the active validator set. Go to the 'unbonded' state.
- `Updated`: Updates (existing) validator data. Does NOT imply entering the active validator set.
- `Jailed`: Validators that leave the active set. Go to the 'jailed' state. Implies slashing.
- `Unjailed`: Validators that enter the validator set.
- `Tombstoned`: Validators that leave the active set permanently. Implies slashing.

Reply:
Good summary. I should add this as Godoc. Some minor details:

> Additions: New and active validators.

This is an addition to the active set. `New` as "first seen" is not fully correct. This can also be a validator that was in the active set before and re-enters it now.

> Updated: Updates (existing) validator data. Does NOT imply entering the active validator set. Does NOT imply validator creation.

This is fully unrelated to the active set and may include any validator in any state. Jailing for example also triggers an `Updated` callback and would be included in this list.

Reply:
Why does jailing triggers an `Updated` callback?

Reply:
The current impl has:

- `Unjailed`: Validators that enter the validator set. Can only come from 'jailed', if not it's ignored.

If `Unjailed` and `Removals` are present, the unjailed validator goes to 'unbonded' instead (i. e. `Removals` has precedence over `Unjailed`).

Finally, `Additions` over an existing validator sets it to 'active' and updates its pubkey. That is, it's like `Updated` but with a state change to 'active'. This is to provide a way for existing validators to enter the active set from 'unbonded'.

If `Tombstoned` is present, it takes precedence over all other state changes, and it's permanent, that is, no other state changes are possible anymore after tombstoning.

Reply:
In the latest version, `Unjailed` has been turned into a no-op, as it cannot be known if it had to change the state to 'active' or to 'unbonded'.
A subsequent or simultaneous addition or removal will do the job.